### DISCUSSION TITLE
Firefox and Chromium don't work

### DISCUSSION THREAD

**Discussion by kolAflash** [kohlschutter/unsock#1] [General]
I'm trying to use unwrap for local website tests. It works for the Apache webserver and the simple [ELinks](https://en.wikipedia.org/wiki/ELinks) browser. But I couldn't get it working with Firefox or Chromium.

Any idea how to fix this?

&nbsp;

Webserver (Debian-12, present working directory as webroot):
```bash
UNSOCK_ADDR=127.0.0.42/32 UNSOCK_DIR="${HOME}"/sockets \
  LD_PRELOAD="${HOME}"/opt/unsock/libunsock.so.1.1.0 \
  apache_mod=/usr/lib/apache2/modules \
  /usr/sbin/apache2 -f /dev/null \
  -C 'LoadModule mpm_event_module "${apache_mod}/mod_mpm_event.so"' \
  -C 'LoadModule authz_core_module "${apache_mod}/mod_authz_core.so"' \
  -C 'Define FOREGROUND' -C 'PidFile "${HOME}/apache-share-8080.pid"' \
  -C 'Listen 127.0.0.42:8080' -C 'DocumentRoot "${PWD}"' \
  -C 'LoadModule mime_module "${apache_mod}/mod_mime.so"' \
  -C 'TypesConfig /etc/mime.types' \
  -C 'ErrorLog "/dev/stderr"' -C 'CustomLog "/dev/stdout" combined' -C 'LogLevel debug'
```

Client:
```bash
UNSOCK_ADDR=127.0.0.42/32 UNSOCK_DIR="${HOME}"/sockets \
  LD_PRELOAD="${HOME}"/opt/unsock/libunsock.so.1.1.0 \
  BROWSER 'http://127.0.0.42:8080/index.html'
```
Replacing Browser with `elinks` (if installed) works.  
But replacing it with `firefox` or `chromium` doesn't work.

Redirecting it back to TCP and accessing it with Firefox works. But that's not really useful.  
first shell: `socat TCP-LISTEN:8080,bind=127.0.0.1,fork UNIX:"${HOME}"/sockets/8080.sock`
second shell: `firefox 'http://127.0.0.1:8080'`

&nbsp;

Naturally I'm assuming, that Firefox and Chromium must not already be running. So new processes get spawned with `LD_PRELOAD` set.

For Firefox the Apache logs a connection. But Firefox is stuck loading and crashes after about a minute.

For Chromium Apache doesn't even log a connection. And finally Chromium reports `ERR_SOCKET_NOT_CONNECTED`.

&nbsp;

P.S.
kohlschutter: Thanks for the great software!

**Comment by kohlschuetter** [comment]
@kolAflash That's interesting.

It would be helpful to `strace` the socket-related system calls when running Firefox or Chromium and see what's going wrong there (I currently neither have an environment to test Linux GUI apps nor the bandwidth to set one up for this)

I could speculate that they create/fork new processes, and then these processes expect some communication to happen, which obviously doesn't work when only one process is being `unsock`-ed.

Have you tried `export`-ing these environment variables instead of invoking only the binary with them? That way, child processes will inherit the configuration:

```
export UNSOCK_ADDR=127.0.0.42/32 UNSOCK_DIR="${HOME}"/sockets
export LD_PRELOAD="${HOME}"/opt/unsock/libunsock.so.1.1.0
"$BROWSER" 'http://127.0.0.42:8080/index.html'
```

**Comment by kolAflash** [comment]
@kohlschuetter  
Thanks for your reply!

As far as I know `export` makes only a difference for the shell from which you start programs.  
`VAR=42 my-program` and `export VAR=42; my-program` looks the same for my-program.  
And all children _(forks)_ of my-program inherit `VAR=42` from my-program.

Nevertheless I tested running Firefox with [the variables](https://github.com/kohlschutter/unsock/discussions/1#discussion-6616507) exported. And the result is the same.
